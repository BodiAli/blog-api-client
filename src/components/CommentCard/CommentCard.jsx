import { useState } from "react";
import { Link } from "react-router";
import { toast } from "react-toastify";
import { formatDistanceToNow } from "date-fns";
import anonymousImage from "../../assets/images/anonymous.jpg";
import optionsIcon from "../../assets/images/dots-vertical.svg";
import editIcon from "../../assets/images/edit.svg";
import deleteIcon from "../../assets/images/delete.svg";
import styles from "./CommentCard.module.css";

export default function CommentCard({ comment, user, token, dispatch }) {
  const [isEditing, setIsEditing] = useState(false);
  const [updateLoading, setUpdateLoading] = useState(false);
  const [invalidInputErrors, setInvalidInputErrors] = useState([]);

  async function handleLikeComment() {
    try {
      const res = await fetch(
        `${import.meta.env.VITE_SERVER_URL}/posts/${comment.postId}/comments/${comment.id}/like`,
        {
          method: "PATCH",
          headers: {
            Authorization: token,
          },
        }
      );

      if (!res.ok) {
        if (comment.commentLiked) {
          throw new Error("Failed to remove like please try again later");
        } else {
          throw new Error("Failed to like comment please try again later");
        }
      }

      if (comment.commentLiked) {
        toast.success("Like removed successfully!");
      } else {
        toast.success("Comment liked successfully!");
      }

      dispatch({ type: "update-comment-like", comment });
    } catch (error) {
      toast.error(error.message);
    }
  }

  async function handleUpdateComment(e) {
    e.preventDefault();

    const formData = new FormData(e.currentTarget);

    const content = formData.get("content");

    try {
      setUpdateLoading(true);
      const res = await fetch(
        `${import.meta.env.VITE_SERVER_URL}/posts/${comment.postId}/comments/${comment.id}`,
        {
          method: "PUT",
          headers: {
            Authorization: token,
            "Content-Type": "application/json",
          },
          body: JSON.stringify({ content }),
        }
      );

      if (!res.ok) {
        if (res.status === 400) {
          const { errors } = await res.json();

          setInvalidInputErrors(errors);
          return;
        }
        throw new Error("Failed to update comment please try again later");
      }

      const { msg } = await res.json();

      dispatch({ type: "update-comment-content", comment, content });

      toast.success(msg);
      setIsEditing(false);
      setInvalidInputErrors([]);
    } catch (error) {
      toast.error(error.message);
      setIsEditing(false);
    } finally {
      setUpdateLoading(false);
    }
  }

  async function handleDeleteComment() {
    try {
      const res = await fetch(
        `${import.meta.env.VITE_SERVER_URL}/posts/${comment.postId}/comments/${comment.id}`,
        {
          method: "DELETE",
          headers: {
            Authorization: token,
          },
        }
      );

      if (!res.ok) {
        throw new Error("Failed to delete comment please try again later");
      }

      dispatch({ type: "delete-comment", comment });
      toast.success("Comment deleted successfully!");
    } catch (error) {
      toast.error(error.message);
    }
  }

  return (
    <div className={styles.card}>
      <div className={styles.container1}>
        <div className={styles.commentUserInfo}>
          <img
            src={comment.User.Profile.profileImgUrl ? comment.User.Profile.profileImgUrl : anonymousImage}
            alt={`${comment.User.firstName} ${comment.User.lastName}'s profile picture`}
          />
          <p className={styles.name}>
            {comment.User.firstName} {comment.User.lastName}
          </p>
          <p className={styles.createdAt}>
            {formatDistanceToNow(comment.updatedAt, { addSuffix: true, includeSeconds: true })}
          </p>
        </div>
        {comment.userId === user.id && (
          <div className={styles.dropdown}>
            <button className={styles.dropBtn}>
              <img src={optionsIcon} alt="options" />
            </button>
            <div className={styles.dropdownContent}>
              <div
                className={styles.folderEdit}
                onClick={() => {
                  setIsEditing(true);
                }}
              >
                <img src={editIcon} alt="Edit comment" />
                <button>Edit</button>
              </div>
              <div onClick={handleDeleteComment} className={styles.folderDelete}>
                <img src={deleteIcon} alt="delete comment" />
                <button>Delete</button>
              </div>
            </div>
          </div>
        )}
      </div>
      <div className={styles.container2}>
        {isEditing ? (
          <form onSubmit={handleUpdateComment} className={styles.editForm}>
            <div>
              <input type="text" name="content" defaultValue={comment.content} required />
              <button
                className={styles.cancel}
                type="button"
                onClick={() => {
                  setInvalidInputErrors([]);
                  setIsEditing(false);
                }}
              >
                Cancel
              </button>
              <button
                className={`${styles.confirm} ${updateLoading ? styles.disabled : ""}`}
                disabled={updateLoading}
                type="submit"
              >
                Confirm
              </button>
            </div>
            <ul className={styles.errors}>
              {invalidInputErrors.map((error, i) => {
                return <li key={i}>{error.msg}</li>;
              })}
            </ul>
          </form>
        ) : (
          <p>{comment.content}</p>
        )}
      </div>
      <div className={styles.likeContainer}>
        <p className={styles.likes}>
          <strong>Likes:</strong> {comment.likes}
        </p>
        {!user ? (
          <p className={styles.loginMessage} data-testid="login-notice-comment">
            Please{" "}
            <Link to="log-in" viewTransition>
              login
            </Link>{" "}
            to be able to like comment
          </p>
        ) : (
          <button
            data-testid="like-button-comment"
            onClick={handleLikeComment}
            className={`${styles.likeButton} ${comment.commentLiked ? styles.liked : ""}`}
          ></button>
        )}
      </div>
    </div>
  );
}
