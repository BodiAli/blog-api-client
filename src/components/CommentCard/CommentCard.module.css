.card {
  padding: 0.5rem;
  display: grid;
  grid-template-columns: 1fr;
  gap: 2rem;
  border-radius: 5px;
  border-bottom: 1px solid rgba(0, 0, 0, 0.514);
}

.textContainer1 {
  display: grid;
  grid-auto-flow: column;
  justify-content: start;
  align-items: center;
  gap: 0.5rem;
}

.textContainer1 > img {
  width: 60px;
  aspect-ratio: 1 / 1;
  object-fit: cover;
  border-radius: 50%;
}

.name {
  font-size: 1.1rem;
}

.createdAt {
  color: #0000009c;
}

.likeContainer {
  display: grid;
  grid-auto-flow: column;
  justify-content: end;
  align-items: center;
  gap: 0.5rem;
}

.likeContainer strong {
  font-weight: 500;
}

.likeButton {
  background-color: black;
  mask-image: url("../../assets/images/heart.svg");
  mask-repeat: no-repeat;
  mask-size: contain;
  width: 25px;
  aspect-ratio: 1 / 1;
  cursor: pointer;
}

.likeButton.liked {
  background-color: crimson;
}

.loginMessage a {
  color: #000;
}
