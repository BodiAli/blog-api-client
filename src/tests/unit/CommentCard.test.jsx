import { screen, render } from "@testing-library/react";
import { createRoutesStub } from "react-router";
import { toast } from "react-toastify";
import userEvent from "@testing-library/user-event";
import { vi, expect, test, beforeEach } from "vitest";
import CommentCard from "../../components/CommentCard/CommentCard";

vi.stubGlobal("fetch", vi.fn());

vi.mock("date-fns", async (importOriginal) => {
  const result = await importOriginal();

  return {
    ...result,
    formatDistanceToNow: () => "1 day ago",
  };
});

vi.mock("react-toastify", () => {
  return {
    toast: {
      success: vi.fn(),
      error: vi.fn(),
    },
  };
});

const fakeComment = {
  id: "7397319e-58ef-4e42-bc94-aa8788db4090",
  userId: 1,
  postId: "57c20dfb-8135-42eb-8abe-0bffbaffd668",
  content: "Comment content",
  likes: 1,
  createdAt: "2025-05-03T07:51:49.273Z",
  updatedAt: "2025-05-04T06:19:03.092Z",
  User: {
    id: 1,
    firstName: "Bodi",
    lastName: "Ali",
    Profile: {
      profileImgUrl: "imageUrl",
    },
  },
  commentLiked: false,
};

const fakeUser = {
  id: 1,
  firstName: "Bodi",
  lastName: "Ali",
  Profile: {
    profileImgUrl: "imageUrl",
  },
};

const fakeToken = "token";

const fakeDispatch = vi.fn();

beforeEach(() => {
  vi.clearAllMocks();
});

describe("CommentCard component", () => {
  test("Should render comment data ", () => {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={fakeUser} />
        ),
      },
    ]);

    const { container } = render(<Stub />);

    expect(container).toMatchSnapshot();
  });

  test("If user is not authenticated component should not render options icon", () => {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={null} />
        ),
      },
    ]);

    render(<Stub />);

    expect(screen.queryByAltText("options")).not.toBeInTheDocument();
  });

  test("If authenticated user is not the author of the comment, component should not render options icon", () => {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard
            comment={fakeComment}
            dispatch={fakeDispatch}
            token={fakeToken}
            user={{ ...fakeUser, id: 2 }}
          />
        ),
      },
    ]);

    render(<Stub />);

    expect(screen.queryByAltText("options")).not.toBeInTheDocument();
  });

  test("If edit button is clicked component should render a form to edit comment", async () => {
    const user = userEvent.setup();

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={fakeUser} />
        ),
      },
    ]);

    render(<Stub />);

    const editButton = screen.getByRole("button", { name: "Edit" });

    await user.click(editButton);

    expect(screen.getByDisplayValue("Comment content")).toBeInTheDocument();
  });

  test("If edit button is clicked and then cancel button is clicked component should hide form", async () => {
    const user = userEvent.setup();

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={fakeUser} />
        ),
      },
    ]);

    render(<Stub />);

    const editButton = screen.getByRole("button", { name: "Edit" });

    await user.click(editButton);

    expect(screen.getByDisplayValue("Comment content")).toBeInTheDocument();

    const cancelButton = screen.getByRole("button", { name: "Cancel" });
    await user.click(cancelButton);

    expect(screen.queryByDisplayValue("Comment content")).not.toBeInTheDocument();
  });

  test("If user is not authenticated component should render a login notice", () => {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={null} />
        ),
      },
    ]);

    render(<Stub />);

    const loginNotice = screen.getByTestId("login-notice-comment");

    expect(loginNotice).toBeInTheDocument();
  });

  test("When like button is clicked component should call dispatch function and a success toast", async () => {
    const user = userEvent.setup();

    fetch.mockImplementationOnce(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
      });
    });

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={fakeUser} />
        ),
      },
    ]);

    render(<Stub />);

    const likeButton = screen.getByTestId("like-button-comment");

    await user.click(likeButton);

    expect(fakeDispatch).toHaveBeenCalledWith({ type: "update-comment-like", comment: fakeComment });
    expect(toast.success).toHaveBeenCalledWith("Comment liked successfully!");
  });

  test("When unlike button is clicked component should call dispatch function and a success toast", async () => {
    const user = userEvent.setup();

    fetch.mockImplementationOnce(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
      });
    });

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard
            comment={{ ...fakeComment, commentLiked: true }}
            dispatch={fakeDispatch}
            token={fakeToken}
            user={fakeUser}
          />
        ),
      },
    ]);

    render(<Stub />);

    const likeButton = screen.getByTestId("like-button-comment");

    await user.click(likeButton);

    expect(fakeDispatch).toHaveBeenCalledWith({
      type: "update-comment-like",
      comment: { ...fakeComment, commentLiked: true },
    });
    expect(toast.success).toHaveBeenCalledWith("Like removed successfully!");
  });

  test("Should toast an error of removing like if comment is liked and fetch is not ok", async () => {
    const user = userEvent.setup();

    fetch.mockImplementationOnce(() => {
      return Promise.resolve({
        ok: false,
      });
    });

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard
            comment={{ ...fakeComment, commentLiked: true }}
            dispatch={fakeDispatch}
            token={fakeToken}
            user={fakeUser}
          />
        ),
      },
    ]);

    render(<Stub />);

    const likeButton = screen.getByTestId("like-button-comment");

    await user.click(likeButton);

    expect(fakeDispatch).not.toHaveBeenCalled();
    expect(toast.error).toHaveBeenCalledWith("Failed to remove like please try again later");
  });

  test("Should toast an error of adding like if comment is not liked and fetch is not ok", async () => {
    const user = userEvent.setup();

    fetch.mockImplementationOnce(() => {
      return Promise.resolve({
        ok: false,
      });
    });

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={fakeUser} />
        ),
      },
    ]);

    render(<Stub />);

    const likeButton = screen.getByTestId("like-button-comment");

    await user.click(likeButton);

    expect(fakeDispatch).not.toHaveBeenCalled();
    expect(toast.error).toHaveBeenCalledWith("Failed to like comment please try again later");
  });

  test("Should show errors sent by the server if update form is sent with invalid input", async () => {
    const user = userEvent.setup();

    fetch.mockImplementationOnce(() => {
      return Promise.resolve({
        ok: false,
        status: 400,
        json: () => Promise.resolve({ errors: [{ msg: "Comment can not be empty." }] }),
      });
    });

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={fakeUser} />
        ),
      },
    ]);

    render(<Stub />);

    await user.click(screen.getByRole("button", { name: "Edit" }));

    const commentInput = screen.getByDisplayValue("Comment content");

    await user.clear(commentInput);
    await user.type(commentInput, "Invalid input");
    await user.click(screen.getByRole("button", { name: "Confirm" }));

    const errorList = screen.getByRole("listitem");
    expect(errorList).toHaveTextContent("Comment can not be empty.");
  });

  test("Should call dispatch, success toast, and hide form when form is successfully submitted", async () => {
    const user = userEvent.setup();

    fetch.mockImplementationOnce(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
        json: () => Promise.resolve({ msg: "Comment updated successfully!" }),
      });
    });

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={fakeUser} />
        ),
      },
    ]);

    render(<Stub />);

    await user.click(screen.getByRole("button", { name: "Edit" }));

    const commentInput = screen.getByDisplayValue("Comment content");

    await user.clear(commentInput);
    await user.type(commentInput, "Comment updated");
    await user.click(screen.getByRole("button", { name: "Confirm" }));

    expect(screen.queryByDisplayValue("Comment updated")).not.toBeInTheDocument();
    expect(toast.success).toBeCalledWith("Comment updated successfully!");
    expect(fakeDispatch).toBeCalledWith({
      type: "update-comment-content",
      comment: fakeComment,
      content: "Comment updated",
    });
  });

  test("Should call dispatch and success toast when delete button is clicked", async () => {
    const user = userEvent.setup();

    fetch.mockImplementationOnce(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
      });
    });

    const Stub = createRoutesStub([
      {
        path: "/",
        Component: () => (
          <CommentCard comment={fakeComment} dispatch={fakeDispatch} token={fakeToken} user={fakeUser} />
        ),
      },
    ]);

    render(<Stub />);

    await user.click(screen.getByRole("button", { name: "Delete" }));

    expect(fakeDispatch).toBeCalledWith({ type: "delete-comment", comment: fakeComment });
    expect(toast.success).toBeCalledWith("Comment deleted successfully!");
  });
});
