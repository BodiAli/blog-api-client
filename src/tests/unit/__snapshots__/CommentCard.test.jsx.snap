// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentCard component > Should render comment data  1`] = `
<div>
  <div
    class="_card_ec5f2a"
  >
    <div
      class="_container1_ec5f2a"
    >
      <div
        class="_commentUserInfo_ec5f2a"
      >
        <img
          alt="Bodi Ali's profile picture"
          src="imageUrl"
        />
        <p
          class="_name_ec5f2a"
        >
          Bodi
           
          Ali
        </p>
        <p
          class="_createdAt_ec5f2a"
        >
          1 day ago
        </p>
      </div>
      <div
        class="_dropdown_ec5f2a"
      >
        <button
          class="_dropBtn_ec5f2a"
        >
          <img
            alt="options"
            src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M12,16A2,2%200%200,1%2014,18A2,2%200%200,1%2012,20A2,2%200%200,1%2010,18A2,2%200%200,1%2012,16M12,10A2,2%200%200,1%2014,12A2,2%200%200,1%2012,14A2,2%200%200,1%2010,12A2,2%200%200,1%2012,10M12,4A2,2%200%200,1%2014,6A2,2%200%200,1%2012,8A2,2%200%200,1%2010,6A2,2%200%200,1%2012,4Z'%20/%3e%3c/svg%3e"
          />
        </button>
        <div
          class="_dropdownContent_ec5f2a"
        >
          <div
            class="_folderEdit_ec5f2a"
          >
            <img
              alt="Edit comment"
              src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3%2017.15,3.1%2016.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65%2021.1,6%2020.71,5.63L18.37,3.29C18.17,3.09%2017.92,3%2017.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z'%20/%3e%3c/svg%3e"
            />
            <button>
              Edit
            </button>
          </div>
          <div
            class="_folderDelete_ec5f2a"
          >
            <img
              alt="delete comment"
              src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M9,3V4H4V6H5V19A2,2%200%200,0%207,21H17A2,2%200%200,0%2019,19V6H20V4H15V3H9M7,6H17V19H7V6M9,8V17H11V8H9M13,8V17H15V8H13Z'%20/%3e%3c/svg%3e"
            />
            <button>
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="_container2_ec5f2a"
    >
      <p>
        Comment content
      </p>
    </div>
    <div
      class="_likeContainer_ec5f2a"
    >
      <p
        class="_likes_ec5f2a"
        data-testid="likes-count-comment"
      >
        <strong>
          Likes:
        </strong>
         
        1
      </p>
      <button
        class="_likeButton_ec5f2a "
        data-testid="like-button-comment"
      />
    </div>
  </div>
</div>
`;
